python/tvm/contrib/stm32/BaseEmitter.py

```diff
@@ -0,0 +1,1257 @@
+    assert strides is None, f"###Error: non-compact tensors are not handled yet."
+    out_c.write(f"AI_STATIC int64_t {dl_tensor_name}_strides[{ndim}] = {{}}; \n")
+
+
+def _emit_tensor_quant(dl_tensor_name, quantization_map, out_c):
+
+    if dl_tensor_name in quantization_map:
+        quantization = quantization_map[dl_tensor_name]
+    else:
+        quantization = None
+
+    if quantization is not None:
+        # minval = quantization['min']
+        # maxval = quantization['max']
+        scale = quantization["scale"]
+        zero_point = quantization["zero_point"]
+        # dim = quantization['dim']
+
+        #
+        # Sometimes we get a scalar with ScaleAsNumpy.
+        # This seem to mean not quantized ?
+        #
+        if not isinstance(scale, np.ndarray):
+            assert scale == 0.0, f"Non-quantized tensor with scale != 0.0"
+            assert (
+                not isinstance(zero_point, np.ndarray) and zero_point == 0
+            ), f"Non-quantized tensor with zero_point != 0"
+            return None
+
+        scale_size = len(scale)
+        zero_point_size = len(zero_point)
+
+        assert len(scale) == len(
+            zero_point
+        ), f"Inconsistent quantizations scale:{scale} vs zero-point:{zero_point}"
+
+        if len(scale) == 1:
+            quant_name = dl_tensor_name + "_quant"
+            out_c.write(f"AI_STATIC float {quant_name}_scale[{scale_size}] = {{ ")
+            for val in scale:
+                out_c.write(f"{val}, ")
+            out_c.write(f"}};\n")
+            out_c.write(f"AI_STATIC int32_t {quant_name}_zero_point[{zero_point_size}] = {{ ")
+            for val in zero_point:
+                out_c.write(f"{val}, ")
+            out_c.write(f"}};\n")
+            out_c.write(f"AI_STATIC ai_quantization_info {quant_name} = {{\n")
+            out_c.write(f"  .scale = {quant_name}_scale,\n")
+            out_c.write(f"  .zero_point = {quant_name}_zero_point,\n")
+            out_c.write(f"  .dim = -1\n")
+            out_c.write(f"}}; \n")
+
+            return quant_name
+
+    return None
+
+
+def emit_tensor_init(dl_tensor_name, tensor, quantization_map, out_c):
+    """Emit tensor definition."""
+    dltype = tensor["dltype"]
+    dims = tensor["dims"]
+    strides = tensor["strides"]
+    if "device" in tensor:
+        device = tensor["device"]
+    else:
+        device = "kDLCPU"
+    # storage_id = tensor['storage_id']
+    byte_offset = tensor["byte_offset"]
+    # size = tensor['size']
+    dtype = _get_type_data(dltype)
+    ndim = len(dims)
+    shape = str(dims)
+    _emit_tensor_shape(dl_tensor_name, ndim, shape, strides, out_c)
+
+    #
+    # Quantization
+    #
+    if quantization_map is not None:
+        quant_name = _emit_tensor_quant(dl_tensor_name, quantization_map, out_c)
+    else:
+        quant_name = None
+
+    #
+    # Contents
+    #
+    out_c.write(f"AI_ALIGNED({8}) AI_STATIC ai_tensor {dl_tensor_name} = {{ \n")
```

Comment:
a suggestion: there are a lot of "templates" like these that might be easier to understand if you write them as python heredoc and take more advantage of the format() function:

```
AI_TENSOR_TEMPLATE = """\
AI_ALIGNED(8) AI_STATIC ai_tensor {dl_tensor_name} = {{
   .dltensor = {{
      .data = (ai_ptr)(NULL),
      .device = {{ {tensor.get("device", "kDLCPU")} }},
      .ndim = {len(tensor["dims"])},
    # etc
```

and then
```
out_c.write(AI_TENSOR_TEMPLATE.format(device_name=device_name, tensor=tensor))
```